src/app/utils/convert-html-to-pdf.ts

```diff
@@ -1,39 +1,218 @@
  * @param summaryHtml HTML to generate PDF from
- * @returns PDF Uint8Array
+ * @returns PDF Buffer
  */
 export const generatePdfFromHtml = async (
   summaryHtml: string,
-): Promise<Uint8Array> => {
-  return tracer.trace('generatePdfFromHtml', async () => {
-    const browser = await puppeteer.launch({
-      args: [
-        '--no-sandbox',
-        '--disable-gpu', // See https://github.com/puppeteer/puppeteer/issues/11640#issuecomment-2361858540
-      ],
-      headless: true,
-      executablePath: config.chromiumBin,
+  isUseLambdaOutput: boolean,
+): Promise<Buffer> => {
+  const logMeta = {
+    action: 'generatePdfFromHtml',
+    isUseLambdaOutput,
+  }
+
+  logger.info({
+    message: 'Generating pdf from html started',
+    meta: logMeta,
+  })
+
+  const localStopwatch = startStopwatch()
+  const localResult = generatePdfFromHtmlLocally(summaryHtml).then((result) => {
+    const latencyMs = localStopwatch.stop()
+    logger.info({
+      message: 'Successfully generated pdf from html using local',
+      meta: { ...logMeta, latencyMs },
     })
-    const page = await browser.newPage()
-    await page.setContent(summaryHtml, {
-      waitUntil: 'networkidle0',
+    submitPdfGenerationLatencyMetric({
+      latencyMs,
+      isLocal: true,
     })
-    const pdfBuffer = await page.pdf({
-      format: 'A4',
-      printBackground: true,
-      margin: {
-        top: '20px',
-        bottom: '40px',
-      },
+    return result
+  })
+
+  const isPdfGenerationLambdaConfigured =
+    !!AwsConfig.pdfGeneratorLambdaFunctionName
+  if (!isPdfGenerationLambdaConfigured) {
+    logger.info({
+      message:
+        'Pdf generation lambda is not configured - using result from local pdf generation',
+      meta: logMeta,
+    })
+    return localResult
+  }
+
+  const lambdaStopwatch = startStopwatch()
+  const lambdaResultAsync = generatePdfFromHtmlLambda(summaryHtml).map(
+    (result) => {
+      const latencyMs = lambdaStopwatch.stop()
+      logger.info({
+        message: 'Successfully generated pdf from html using lambda',
+        meta: { ...logMeta, latencyMs },
+      })
+      submitPdfGenerationLatencyMetric({
+        latencyMs,
+        isLocal: false,
+      })
+      return result
+    },
+  )
+
+  if (isUseLambdaOutput) {
+    const lambdaResult = await lambdaResultAsync
+    if (lambdaResult.isErr()) {
+      logger.error({
+        message: 'Error generating pdf from html using lambda',
+        meta: logMeta,
+        error: lambdaResult.error,
+      })
+      throw lambdaResult.error
+    }
+
+    logger.info({
+      message:
+        'Successfully generated pdf from html - using result from lambda pdf generation',
+      meta: logMeta,
     })
-    await browser.close()
-    return pdfBuffer
+    return lambdaResult.value
+  }
+
+  logger.info({
+    message:
+      'Successfully generated pdf from html - using result from local pdf generation',
+    meta: logMeta,
   })
+  return await localResult
```

Comment:
thanks @kevin9foong!

just a few small nits on the code structure here:
- this code works and is achieves the goal of getting our hypothesis tested early, but we might have to clean this up later.
- the final return for the local generation codepath is below the lambda generation codepath
- the return in the lambda generation codepath immediately returns the top-level function`generatePdfHtml` itself, which behaves differently from the local generation codepath

it could lead to a footgun later on, since the differing coding styles (return a promise vs immediately return) is a code readability issue.